### PR TITLE
[fully_async, trainer] fix: sync optimizer total steps before trainer initialization

### PR DESCRIPTION
**Pull request by mikequan0425**
### What does this PR do?

> Add **concise** overview of what this PR aims to achieve or accomplish. Reference related GitHub issues and PRs that help with the review.

It was observed in experiments that the learning rate (lr) was always 0. This issue does not occur when starting the script via main_ppo, and only emerges under fully async mode. Refer to the following for detailed problem description:https://github.com/verl-project/verl/issues/6683

In short, optim.total_training_steps is not assigned correctly during the initialization of trainer optim.

image

### Design & Code Changes

> Demonstrate the high-level design if this PR is complex, and list the specific changes.

The simplest approach is to pass values through by configuring `actor_rollout_ref.actor.optim.total_training_steps`, yet the actual trainer step should be calculated as total_rollout_steps / (required_samples * trigger_parameter_sync_step)

Therefore, this PR provides a method to assign the corresponding value to optim.total_training_steps when creating the trainer.

### Checklist Before Starting

- [x] Search for similar PRs. Paste at least one query link here: https://github.com/verl-project/verl/pulls?q=is%3Apr+is%3Aopen+fully+async+lr
- [x] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `fsdp`, `megatron`, `veomni`, `sglang`, `vllm`, `rollout`, `trainer`, `ci`, `training_utils`, `recipe`, `hardware`, `deployment`, `ray`, `worker`, `single_controller`, `misc`, `perf`, `model`, `algo`, `env`, `tool`, `ckpt`, `doc`, `data`, `cfg`, `reward`, `fully_async`, `one_step_off`
  - If this PR involves multiple modules, separate them with `,` like `[megatron, fsdp, doc]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][fsdp, megatron] feat: dynamic batching`

### Test

> For changes that can not be tested by CI (e.g., algorithm implementation, new model support), validate by experiment(s) and show results like training curve plots, evaluation results, etc.

image

### API and Usage Example

> Demonstrate how the API changes if any, and provide usage example(s) if possible.

```python
# Add code snippet or script demonstrating how to use this
```

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [ ] Read the [Contribute Guide](https://github.com/verl-project/verl/blob/main/CONTRIBUTING.md).
- [ ] Apply [pre-commit checks](https://github.com/verl-project/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [ ] Add / Update [the documentation](https://github.com/verl-project/verl/tree/main/docs).
- [ ] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/verl-project/verl/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...
- [ ] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1) in [the `verl` Slack workspace](https://join.slack.com/t/verl-project/shared_invite/zt-3855yhg8g-CTkqXu~hKojPCmo7k_yXTQ). (If not accessible, please try [the Feishu group (飞书群)](https://applink.larkoffice.com/client/chat/chatter/add_by_link?link_token=772jd4f1-cd91-441e-a820-498c6614126a).)
- [ ] If your PR is related to the `recipe` submodule, please also update the reference to the submodule commit via `git submodule update --remote` or `cd recipe && git pull origin main`.

### CODE REVIEW

File: verl/experimental/fully_async_policy/fully_async_trainer.py [hunk 1]
```diff
@@ -266,7 +271,16 @@ def set_total_train_steps(self, total_training_steps):
...
 
-        self.progress_bar = tqdm(total=self.total_train_steps, initial=0, desc="Training Progress")
+    def _resolve_total_training_steps_before_init(self):
+        rollout_total_steps = self.config.rollout.total_rollout_steps
+        if rollout_total_steps is None:
+            return None
+
+        required_samples = (
+            self.config.actor_rollout_ref.actor.ppo_mini_batch_size * self.config.async_training.require_batches
+        )
```
Comment (Luosuu) [inline verl/experimental/fully_async_policy/fully_async_trainer.py]: `optim.total_training_steps` is consumed by the LR scheduler, and the scheduler is stepped on every actor update (`_fit_update_actor` -> `update_actor`), not only when parameters are synced. Dividing by `trigger_parameter_sync_step` makes the schedule finish `trigger_parameter_sync_step` times too early; if the progress bar/checkpoint version wants sync steps, please keep that separate from the optimizer scheduler steps.

File: verl/experimental/fully_async_policy/fully_async_trainer.py [hunk 2]
```diff
@@ -266,7 +271,16 @@ def set_total_train_steps(self, total_training_steps):
         except Exception as e:
             print(f"Warning: Could not set total_training_steps in config. Structure missing? Error: {e}")
```
Comment (Luosuu) [inline verl/experimental/fully_async_policy/fully_async_trainer.py]: This uses the raw configured `rollout.total_rollout_steps`, but the rollouter later computes the effective count as `min(config.rollout.total_rollout_steps, len(train_dataloader) * total_epochs)` and also supports `None`. Since the optimizer is already constructed during `init_workers`, the later `set_total_train_steps()` call cannot rebuild the scheduler, so dataset-limited or unset configs still get the wrong LR schedule here.

**Review by Luosuu — COMMENTED** [review]
Found two issues in the pre-init total step calculation that should be addressed before this is safe.

File: verl/experimental/fully_async_policy/fully_async_trainer.py (hunk 2 above)
Comment (tardis-key) [reply verl/experimental/fully_async_policy/fully_async_trainer.py]: I think the best solution is to finish initializing the rollout part first, and then have the trainer use the rollout_step that was set by rollout for subsequent calculations. Otherwise the total-step computation logic for rollout would need to be maintained in multiple places, which makes further evolution difficult.

File: verl/experimental/fully_async_policy/fully_async_trainer.py (hunk 1 above)
Comment (mikequan0425) [reply verl/experimental/fully_async_policy/fully_async_trainer.py]: Thanks for your review. I originally ported it directly from the set_total_train_steps method in rollouter, and I will fix it.

File: verl/experimental/fully_async_policy/fully_async_trainer.py (hunk 2 above)
Comment (mikequan0425) [reply verl/experimental/fully_async_policy/fully_async_trainer.py]: I agree with both of you. But we cannot simply move create_rollouter before create_trainer. This is because actor_wg only exists within the trainer after trainer.init_workers() is called, and rollouter.init_workers() will read this hybrid worker group during its initialization process. Meanwhile, we can move the logic for sync total_train_steps between the rollouter and the trainer to before create_trainer, and remove the original corresponding code.

**Review by tardis-key — CHANGES_REQUESTED** [review]

File: verl/experimental/fully_async_policy/fully_async_main.py [hunk 3]
```diff
@@ -114,14 +119,39 @@ def _initialize_components(self, config) -> None:
 
         print("[ASYNC MAIN] All components initialized successfully")
 
-    def _create_rollouter(self, config) -> None:
-        print("[ASYNC MAIN] Starting create rollouter...")
+    def _set_total_training_steps_for_optimizer(self, config, optim_total_train_steps):
```
Comment (tardis-key) [inline verl/experimental/fully_async_policy/fully_async_main.py]: don't move this func to main, it should belong to Trainer. You can change the func to class/static func, then you can call it before initing Trainer

**Review by tardis-key — APPROVED** [review]
@Luosuu lgtm, please review

**Review by Luosuu — APPROVED** [review]

**Comment by Luosuu** [comment]
@mikequan0425 please fix pre-commit

**Comment by wuxibin89** [comment]
The verl/experimental/fully_async_policy is in maintenance mode, we encourage user to use new V1 trainer:
```bash
trainer.use_v1=True
trainer.v1.trainer_mode=colocate_async # sync/colocate_async/separate_async
```
